_posts/2019-12-18-road-to-scala-3.md

```diff
@@ -0,0 +1,174 @@
+
+### What is TASTy?
+
+The Scala 3 compiler serializes every program that it compiles in a compact representation called TASTy. The name TASTy stands for "Typed Abstract Syntax Trees (y)", which hints at the representation format: it is the full program code in the form of syntax trees, which is what the compiler uses internally, and it's fully typed. This means that names are resolved, types are inferred and checked, overloads are resolved, and implicits are inserted. All of this information is serialized.
+
+Today, TASTy is used for separate compilation: when a source file uses `Foo.method` where `Foo` is a class on the classpath, the compiler will read the TASTy of `Foo` in order to know what is the type of `method`. Looking at the class file `Foo.class` would not work because Scala-specific features such as union types cannot be represented in class files. Scala 2 uses a very similar mechanism, but instead of serializing the full program code, it only serializes the types, which is sufficient for separate compilation.
+
+In the future, TASTy mitigates "accidental" binary incompatibilities caused by the way Scala-specific features must be encoded in JVM class files. Because TASTy precisely captures the result of the frontend (after the typechecker has fully qualified all names, inferred types and implicits, resolved overloads, expanded syntactic sugar, etc), the Scala compiler backend can be rerun automatically to derive new class files from TASTy serialized by a previous run. It's a promising compromise between the brittleness of bytecode and the complexity of recompiling from source. See also the Q&A on binary compatibility.
+
+TASTy will also serve as a basis for new or updated tooling in the future (code analysis, IDE support, generating documentation).
+
+### How is binary compatibility handled in Scala 3?
+
+Evolving Scala libraries in a binary compatible fashion is difficult because Scala features have to be encoded in JVM class files. In Java, there's a direct correspondence between source and binary compatibility. TASTy aims to bring the same to Scala, without developing a Scala specific VM, by determining compatibility at the level of (fully-typed) sources, rather than compiled binaries. Currently, the nature of the class file encoding means that backwards compatible changes at the source level, such as adding a field to a trait, actually break binary compatibility.
+
+It's important to note that TASTy is not a silver bullet: it cannot fix breaking changes to APIs – those can only be resolved by (potentially automated) refactorings.
+
+By re-compiling an existing library from its TASTy, its class files are regenerated to work correctly with a new version of its dependencies. For example, a library that was compiled against the Scala 3.2 standard library can be safely used with Scala 3.4. There is no need for library maintainers to re-publish when a new Scala 3.x minor release becomes available.
```

Comment:
It would be good to mention once again sjsir files and nir files in this section, in addition to class files.

Reply:
sorry, I think the blog is already too long